testing/base/source_gen.cpp

```diff
@@ -0,0 +1,571 @@
+    -> void {
+  auto& [count, unique_ids] = unique_ids_by_length.Insert(length, {}).value();
+
+  if (count < number) {
+    unique_ids.GrowForInsertCount(count - number);
+    for ([[maybe_unused]] int i : llvm::seq<int>(count, number)) {
+      char* id_storage = reinterpret_cast<char*>(
+          storage.Allocate(/*Size=*/length, /*Alignment=*/1));
+      for (;;) {
+        std::string new_id_tmp = GenerateRandomIdentifier(length);
+        memcpy(id_storage, new_id_tmp.data(), length);
+        llvm::StringRef new_id(id_storage, length);
+        auto result = unique_ids.Insert(new_id);
+        if (result.is_inserted()) {
+          break;
+        }
+      }
+    }
+    count = number;
+  }
+  unique_ids.ForEach([&](llvm::StringRef id) {
+    if (number > 0) {
+      dest.push_back(id);
+      --number;
+    }
+  });
+  CARBON_CHECK(number == 0);
+}
+
+auto SourceGen::GetShuffledInts(int number, int min, int max)
+    -> llvm::SmallVector<int> {
+  llvm::SmallVector<int> ints;
+  ints.reserve(number);
+
+  // Evenly distribute to each value between min and max.
+  int num_values = max - min + 1;
+  for (int i : llvm::seq_inclusive(min, max)) {
+    int i_count = number / num_values;
+    i_count += i < (min + (number % num_values));
+    ints.append(i_count, i);
+  }
+  CARBON_CHECK(static_cast<int>(ints.size()) == number);
+
+  std::shuffle(ints.begin(), ints.end(), rng);
+  return ints;
+}
+
+auto SourceGen::GetClassGenState(int number, ClassParams params)
+    -> ClassGenState {
+  ClassGenState state;
+  state.public_function_param_counts =
+      GetShuffledInts(number * params.public_function_decls, 0,
+                      params.public_function_decl_params.max_params);
+  state.public_method_param_counts =
+      GetShuffledInts(number * params.public_method_decls, 0,
+                      params.public_method_decl_params.max_params);
+  state.private_function_param_counts =
+      GetShuffledInts(number * params.private_function_decls, 0,
+                      params.private_function_decl_params.max_params);
+  state.private_method_param_counts =
+      GetShuffledInts(number * params.private_method_decls, 0,
+                      params.private_method_decl_params.max_params);
+
+  state.class_names = GetShuffledUniqueIds(number, /*min_length=*/5);
+  int num_members =
+      number * (params.public_function_decls + params.public_method_decls +
+                params.private_function_decls + params.private_method_decls +
+                params.private_field_decls);
+  state.member_names = GetShuffledIds(num_members, /*min_length=*/4);
+  int num_params = Sum(state.public_function_param_counts) +
+                   Sum(state.public_method_param_counts) +
+                   Sum(state.private_function_param_counts) +
+                   Sum(state.private_method_param_counts);
+  state.param_names = GetShuffledIds(num_params);
+  return state;
+}
+
+class SourceGen::UniqueIdPopper {
```

Comment:
This class doesn't have comments in the header, nor on its members.

Reply:
Added, sorry about that.